### PR TITLE
Add FXIOS-11174 Add ActivityAttributes struct to define static and dynamic data for the Live Activity

### PR DESCRIPTION
**Pull request by RyanL123**
## :scroll: Tickets
[Jira ticket](https://mozilla-hub.atlassian.net/browse/FXIOS-11174)
[Github issue](https://github.com/mozilla-mobile/firefox-ios/issues/24333)

## :bulb: Description

This PR adds static and dynamic data to the ActivityAttrbutes struct for the Live Activity showing downloads progress. 

To support multiple downloads (https://github.com/mozilla-mobile/firefox-ios/issues/24374), each entry is stored separately and assigned a unique UUID in order to be updated. It is up to the caller to assign a UUID.

To support https://github.com/mozilla-mobile/firefox-ios/issues/24375, downloads with `hasContentEncoding` marked as `true` will not be counted as part of the download progress.

A lot of attributes are shared with the `Download` class in `DownloadQueue.swift`. To simplify the implementation, the struct here serves as a simple wrapper of the core attributes that we care about in the live activity. It is up to the caller that updates the live activity to identify the correct item to update.

Consumers of the attribute can monitor the overall download progress with the provided methods on the struct.

## :pencil: Checklist
You have to check all boxes before merging
- [x] Filled in the above information (tickets numbers and description of your work)
- [x] Updated the PR name to follow our [PR naming guidelines](https://github.com/mozilla-mobile/firefox-ios/wiki/Pull-Request-Naming-Guide)
- [ ] Wrote unit tests and/or ensured the tests suite is passing
- [ ] When working on UI, I checked and implemented accessibility (minimum Dynamic Text and VoiceOver)
- [ ] If needed, I updated documentation / comments for complex code and public methods
- [ ] If needed, added a backport comment (example `@Mergifyio backport release/v120`)

### CODE REVIEW

**Comment by RyanL123** [comment]
Wondering if unit tests are needed at this point? Or should tests be written along with the code that initiates the live activity?

File: firefox-ios/WidgetKit/DownloadManager/DownloadLiveActivity.swift [hunk 1]
```diff
@@ -5,9 +5,52 @@
...
+            var hasContentEncoding: Bool?
+            var downloadPath: URL
+
+            var totalBytesExpected: Int64?
+            var bytesDownloaded: UInt64
+            var isComplete: Bool
+        }
+        var downloads: [Download]
+
+        func getCompletedDownloads() -> Int {
```
Comment (ih-codes) [inline firefox-ios/WidgetKit/DownloadManager/DownloadLiveActivity.swift:25]:
```suggestion
        var completedDownloads: Int {
```
I suggest we make this a simple computed variable in Swift since it's just a value.

File: firefox-ios/WidgetKit/DownloadManager/DownloadLiveActivity.swift [hunk 2]
```diff
@@ -5,9 +5,52 @@
...
+            return downloads.filter { $0.isComplete }.count
+        }
+
+        func getTotalDownloads() -> Int {
+            return downloads.count
+        }
```
Comment (ih-codes) [inline firefox-ios/WidgetKit/DownloadManager/DownloadLiveActivity.swift:31]: Along the same lines as my previous comment, this could be a computed variable as well.

File: firefox-ios/WidgetKit/DownloadManager/DownloadLiveActivity.swift [hunk 3]
```diff
@@ -5,9 +5,52 @@
...
+
+        func getTotalProgress() -> Double {
+            var totalBytesExpected: UInt64 = 0
+            var totalBytesDownloaded: UInt64 = 0
+
+            for download in downloads {
+                // downloads with content encoding cannot
+                // be estimated accurately and should be
+                // skipped entirely in the calculation of progress
```
Comment (ih-codes) [inline firefox-ios/WidgetKit/DownloadManager/DownloadLiveActivity.swift:40]: Nitpick: We generally try to write our comments to the full line width accepted by our swiftlint rules.

It would be great if you could configure your Xcode to set the page guide to column 125 (which is the max line length our swiftlint linter accepts). Then you can ensure Xcode is set to use a monospaced font for documentation and comments. This will give you a visual ruler so you can write your comments up to column 125, rather than multi-line them unnecessarily.

I'm going to attach a slide from our team onboarding documentation with instructions on how to configure this:
Screenshot 2025-02-20 at 1 06 58 PM

**Review by ih-codes — COMMENTED** [review]
Hi @RyanL123 , thanks a bunch for your detailed PR description, it's very much appreciated! 👏  And great work on this PR. I only have a few very small nitpicks.

> Wondering if unit tests are needed at this point? Or should tests be written along with the code that initiates the live activity?

I would love to see unit tests added as code is committed. In this case I think we would benefit from a few simple unit tests around the new `Download` struct testing edge cases (e.g. what is the expected progress output when no downloads have been started, etc.). These tests will be quite trivial for now but can be rounded out more as new functionality is added later.

If in the future there's a task where the unit tests are going to be a lot of work, let me know and we can make a separate ticket for those.

File: firefox-ios/WidgetKit/DownloadManager/DownloadLiveActivity.swift [hunk 4]
```diff
@@ -5,9 +5,52 @@
...
 import ActivityKit
 import SwiftUI
+import Foundation
 
 struct DownloadLiveActivityAttributes: ActivityAttributes {
     struct ContentState: Codable, Hashable {
+        struct Download: Codable, Hashable {
+            var id: UUID
+            var fileName: String
+            var mimeType: String
```
Comment (DanielDervishi) [inline firefox-ios/WidgetKit/DownloadManager/DownloadLiveActivity.swift:15]: I think we can leave out the mimeType (unless there's a good reason to keep it)

File: firefox-ios/WidgetKit/DownloadManager/DownloadLiveActivity.swift [hunk 5]
```diff
@@ -5,9 +5,52 @@
...
+            var hasContentEncoding: Bool?
+            var downloadPath: URL
+
+            var totalBytesExpected: Int64?
+            var bytesDownloaded: UInt64
```
Comment (DanielDervishi) [inline firefox-ios/WidgetKit/DownloadManager/DownloadLiveActivity.swift:20]: Lets make this Int64 instead to better match the Download class in client (since we'll be converting instances of that class to this struct)

File: firefox-ios/WidgetKit/DownloadManager/DownloadLiveActivity.swift [hunk 6]
```diff
@@ -5,9 +5,51 @@
...
+
+        var completedDownloads: Int {
+            return downloads.filter { $0.isComplete }.count
+        }
+
+        var totalDownloads: Int {
+            return downloads.count
+        }
+
+        func getTotalProgress() -> Double {
```
Comment (DanielDervishi) [inline firefox-ios/WidgetKit/DownloadManager/DownloadLiveActivity.swift:33]: Might be helpful to make totalBytesExpected and totalBytesDownloaded computed variables. Since the live activity will be using basically what totalBytesExpected and totalBytesDownloaded are to show the progress e.g.("10MB of 20MB") it would probably be helpful to include the two as computed properties. This way the UI element can more easily access these values.

**Review by DanielDervishi — COMMENTED** [review]
Hey Ryan, great work! I just left a couple of small comments :)

File: firefox-ios/WidgetKit/DownloadManager/DownloadLiveActivity.swift (hunk 3 above)
Comment (RyanL123) [reply firefox-ios/WidgetKit/DownloadManager/DownloadLiveActivity.swift]: Got it! FYI it's now called "Show reformatting guide" under the Editing section as of Xcode 16.2 if the team wants to update the guide.

File: firefox-ios/WidgetKit/DownloadManager/DownloadLiveActivity.swift [hunk 7]
```diff
@@ -5,9 +5,52 @@
...
+import Foundation
 
 struct DownloadLiveActivityAttributes: ActivityAttributes {
     struct ContentState: Codable, Hashable {
+        struct Download: Codable, Hashable {
+            var id: UUID
+            var fileName: String
+            var mimeType: String
+            var hasContentEncoding: Bool?
+            var downloadPath: URL
```
[the remark this answers is not in the document]
Comment (RyanL123) [reply firefox-ios/WidgetKit/DownloadManager/DownloadLiveActivity.swift:17]: I saw a ticket talking about deep linking, but its only about deep linking to the downloads panel rather than individual files so yeah this shouldn't be needed

File: firefox-ios/WidgetKit/DownloadManager/DownloadLiveActivity.swift (hunk 3 above)
Comment (ih-codes) [reply firefox-ios/WidgetKit/DownloadManager/DownloadLiveActivity.swift]: Haha wow, funny how fast settings change! Thanks Ryan, I'm going to go update this related doc right now. 👌

File: firefox-ios/WidgetKit/DownloadManager/DownloadLiveActivity.swift [hunk 8]
```diff
@@ -5,9 +5,51 @@
...
+
+            var totalBytesExpected: Int64?
+            var bytesDownloaded: UInt64
+            var isComplete: Bool
```
Comment (ih-codes) [inline firefox-ios/WidgetKit/DownloadManager/DownloadLiveActivity.swift:21]: On further thought, we may want to simply remove the download from the array once it is complete. That way only active downloads are ever in the array. 

What do you think @DanielDervishi ? I'm not 100% on this but I think that may be what the existing app download manager does. 🤔

File: firefox-ios/firefox-ios-tests/Tests/ClientTests/DownloadLiveActivityTest.swift [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+        )
+
+        // download 1 and download 4 progress should be ignored
+        XCTAssertEqual(contentState.getTotalProgress(), 0)
+    }
+
+    func testGetTotalProgressWhenNoDownloads() {
+        let contentState = DownloadLiveActivityAttributes.ContentState(downloads: [])
+
+        XCTAssertEqual(contentState.getTotalProgress(), 0)
```
Comment (ih-codes) [inline firefox-ios/firefox-ios-tests/Tests/ClientTests/DownloadLiveActivityTest.swift:83]: You can check the computed properties here too to make sure they are what you expect.

**Review by ih-codes — COMMENTED** [review]
Thanks for adding the new unit tests @RyanL123, and good review @DanielDervishi 👍 

Here are a few more thoughts for your consideration. I'll defer to @DanielDervishi for the final review and approval. 😃

File: firefox-ios/WidgetKit/DownloadManager/DownloadLiveActivity.swift (hunk 8 above)
Comment (DanielDervishi) [reply firefox-ios/WidgetKit/DownloadManager/DownloadLiveActivity.swift:19]: Hey @ih-codes, that's definitely an option that we could go with. I think that may conflict with our current designs and may require us to re-think how we would handle showing information to users about multiple downloads. Like for example, if we continue showing total progress instead of switching to individual progress, we will see the total progress and download progress so far dip whenever a download is complete. We could decide to show download progress sequentially in this case? For example, displaying the following information:  "Download 1 of 3 12MB/24MB", where after the first download finished we show something like "Download 1 of 2 12MB/24MB", Although I'm not totally sure if downloads are handled concurrently or sequentially. If they are handled concurrently, the updates to the information displayed in the live activity may be a bit confusing to the user.

Let me know your thoughts!

Comment (ih-codes) [reply firefox-ios/WidgetKit/DownloadManager/DownloadLiveActivity.swift:19]: All excellent points @DanielDervishi!

We can handle the downloads concurrently. It was one of the reasons I had our mockups say "Downloading X files" instead of a "1 of X", because the latter implies sequential downloading. Which is something I find confusing with the current download manager. 🤔 

For the Live Activities, I was personally imagining that the progress amounts would update as the downloads complete (even though there will be a jump in the circular progress indicator). But, both ways you describe are reasonable approaches. I'm totally fine if you have already designed this to show the total download count until all downloads complete. Please continue on with your plan. 🙂 

Here is one edge case I was considering:
Last fall when I tested downloading a bunch of small files and one really large file, the app's progress toast visually suggests none of the small downloads ever complete (continues to say `1 of N files`) until the big download also completes. But in reality, the small files all finish quick and it's the one big one that goes forever. Our current UI is therefore a bit confusing when you "cancel" the download. The other files have already completed so you can't, in fact, cancel them.

Also, this doesn't even get into the fact that if you download the small file first, then the big one, it continues to display the name of the small file even though it has finished long ago and the larger one is still downloading... 😅

File: firefox-ios/firefox-ios-tests/Tests/ClientTests/DownloadLiveActivityTest.swift [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+        let download2 = makeDownload(type: DownloadType.contentEncoded, isComplete: true)
+        let download3 = makeDownload(type: DownloadType.nilExpectedBytes, isComplete: false)
+        let download4 = makeDownload(type: DownloadType.normal, totalBytesExpected: nil, isComplete: false)
+        let contentState = DownloadLiveActivityAttributes.ContentState(
+            downloads: [download1, download2, download3, download4]
+        )
+
+        XCTAssertEqual(contentState.completedDownloads, 2)
+        XCTAssertEqual(contentState.totalDownloads, 4)
+    }
```
Comment (ih-codes) [inline firefox-ios/firefox-ios-tests/Tests/ClientTests/DownloadLiveActivityTest.swift:22]: I'm leaving the final review up to @DanielDervishi, but I just wanted to say GREAT job on these tests @RyanL123! Excellent work. 👏

File: firefox-ios/WidgetKit/DownloadManager/DownloadLiveActivity.swift (hunk 8 above)
Comment (DanielDervishi) [reply firefox-ios/WidgetKit/DownloadManager/DownloadLiveActivity.swift:19]: Hey Isabella, good point on the edge case. We could maybe also display the number of completed files to better communicate that some of the downloads have completed since we have access to that information.

I think it's best to stick with what Ryan has at the moment when it comes to how we represent state in the download live activity so we can mimic the behaviour of our toast for now :)

Comment (ih-codes) [reply firefox-ios/WidgetKit/DownloadManager/DownloadLiveActivity.swift:19]: Sounds good to me @DanielDervishi! We don't want to hold Ryan up.

**Review by DanielDervishi — COMMENTED** [review]
Great work Ryan, just left one comment :)

**Review by ih-codes — APPROVED** [review]
